01_stratified_classification/nbconverted/lasso_range_gene_learning_rate.py

```diff
@@ -189,15 +214,99 @@
 
+if output_plots:
+    plt.savefig(os.path.join(output_plots_dir, f'{plot_gene}_learning_rate_schedule_coefs.svg'), bbox_inches='tight')
+
+
+# ### Plot loss values
+# 
+# We want to separate the log-likelihood loss (data loss) from the weight penalty (regularization term) in the logistic regression loss function, to see if that breakdown is any different between optimizers.
+
+# In[8]:
+
+
+# get loss function values from file
+def get_loss_values(results_dir, lr_schedule):
+    loss_df = pd.DataFrame()
+    for gene_name in os.listdir(results_dir):
+        if plot_gene not in gene_name: continue
+        gene_dir = os.path.join(results_dir, gene_name)
+        if not os.path.isdir(gene_dir): continue
+        for results_file in os.listdir(gene_dir):
+            if not 'loss_values' in results_file: continue
+            if results_file[0] == '.': continue
+            full_results_file = os.path.join(gene_dir, results_file)
+            gene_loss_df = pd.read_csv(full_results_file, sep='\t', index_col=0)
+            seed = int(results_file.split('_')[-5].replace('s', ''))
+            lasso_param = float(results_file.split('_')[-3].replace('c', ''))
+            gene_loss_df['lasso_param'] = lasso_param
+            gene_loss_df['lr_schedule'] = lr_schedule
+            loss_df = pd.concat((loss_df, gene_loss_df))
+    return loss_df.reset_index(drop=True)
+
+
+# In[9]:
+
+
+all_loss_df = []
+
+for lr_schedule in lr_schedules:
+    all_loss_df.append(
+        get_loss_values(results_dirs[lr_schedule], lr_schedule)
+    )
+    
+all_loss_df = pd.concat(all_loss_df)
+all_loss_df['total_loss'] = all_loss_df.log_loss + all_loss_df.l1_penalty
+all_loss_df = all_loss_df.melt(
+    id_vars=['lr_schedule', 'seed', 'fold', 'lasso_param'],
+    var_name='loss_component',
+    value_name='loss_value'
+)
+
+print(all_loss_df.lr_schedule.unique())
+print(all_loss_df.shape)
+all_loss_df.head()
+
+
+# In[10]:
+
+
+sns.set_style('ticks')
+
+with sns.plotting_context('notebook', font_scale=1.6):
+    g = sns.relplot(
+        data=all_loss_df[all_loss_df.lr_schedule != 'constant_search'].sort_values(by='lr_schedule'),
+        x='lasso_param', y='loss_value', hue='loss_component',
+        hue_order=['log_loss', 'l1_penalty', 'total_loss'],
+        marker='o', kind='line', col='lr_schedule',
+        col_wrap=2, height=5, aspect=1.6,
+        facet_kws={'sharex': False}
+    )
+    g.set(xscale='log', yscale='log', xlim=(10e-8, 10e2), ylim=(10e-18, 10e3))
+    g.set_xlabels('SGD LASSO parameter (lower = less regularization)')
+    for ax in g.axes.flat:
+        ax.set_ylabel('Loss value', visible=True)
+        ax.tick_params(labelleft=True)
+    g.set_titles('Learning rate schedule: {col_name}')
+    sns.move_legend(g, "center", bbox_to_anchor=[1.075, 0.5], frameon=True)
+    g._legend.set_title('Loss component')
+     
+    plt.suptitle(f'LASSO parameter vs. training loss, {plot_gene}', y=1.01)
+
+plt.tight_layout(w_pad=8, h_pad=5)
```

Comment:
What's the significance of the vertical orange line? Also why is it partially shaded on the optimal one?

Reply:
Those are places in the plots where the loss function is 0 (and on the optimal one the shading is a confidence interval since some CV folds are 0 and others aren't). I think it makes more sense visually to just have it be a very small number (and explain it in the figure caption) instead of log(0) = -infinity, so I'll make that change.